### PR TITLE
Option to stop robots inc banners views

### PR DESCRIPTION
**Pull request by jurihahn**
Pull Request for Issue # .

### Summary of Changes
Option to stop increment count of banner views if search engine is detected. Count views as before by default for BC.

### Testing Instructions
Coder review
OR
1. Create banner and open page with this banner. Check if banner view count increment.
2. Set option "Robots Increment Impressions" to "NO"
3. Reload page with banner. Check if banner impressions count increment. Must be +1.
4. Change user-agent to anything includes "Googlebot".
5. Reload page with banner. Check if banner impressions count increment. Must be same as before.
PS: Its can be tested local too. For example in Firefox you can open Developer tools and start Responsive/Mobile view, then add new Device with any user agent.

### Expected result
If set option "Robots increment impressions" to "NO", Module should not increment banner views if search engine detected

### Actual result
Module increments count of banner impressions always

### Documentation Changes Required
No?

### CODE REVIEW

File: language/en-GB/en-GB.mod_banners.ini [hunk 1]
```diff
@@ -26,3 +26,5 @@ MOD_BANNERS_FIELD_TARGET_LABEL="Target"
 MOD_BANNERS_VALUE_STICKYORDERING="Pinned, Ordering"
 MOD_BANNERS_VALUE_STICKYRANDOMISE="Pinned, Randomise"
 MOD_BANNERS_XML_DESCRIPTION="The Banner Module displays the active Banners from the Component."
+MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_LABEL="Robots increment views"
```
Comment (Quy) [inline language/en-GB/en-GB.mod_banners.ini]: Sort keys in alpha order.

File: language/en-GB/en-GB.mod_banners.ini [hunk 2]
```diff
@@ -19,6 +19,8 @@ MOD_BANNERS_FIELD_HEADER_DESC="Text or HTML to display before the group of banne
 MOD_BANNERS_FIELD_HEADER_LABEL="Header Text"
 MOD_BANNERS_FIELD_RANDOMISE_DESC="Randomise the ordering of the banners."
 MOD_BANNERS_FIELD_RANDOMISE_LABEL="Randomise"
+MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_LABEL="Robots increment views"
```
Comment (Quy) [inline language/en-GB/en-GB.mod_banners.ini]: Capitalize words `Robots Increment Views`

Comment (Quy) [reply language/en-GB/en-GB.mod_banners.ini]: Should it be `Impressions`?

File: language/en-GB/en-GB.mod_banners.ini [hunk 3]
```diff
@@ -19,8 +19,8 @@ MOD_BANNERS_FIELD_HEADER_DESC="Text or HTML to display before the group of banne
 MOD_BANNERS_FIELD_HEADER_LABEL="Header Text"
 MOD_BANNERS_FIELD_RANDOMISE_DESC="Randomise the ordering of the banners."
 MOD_BANNERS_FIELD_RANDOMISE_LABEL="Randomise"
-MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_LABEL="Robots increment views"
-MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_DESC="Search engines increase the number of banner views"
+MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_LABEL="Robots Increment Impressions"
+MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_DESC="Search engines increase the number of banner Impressions"
```
Comment (Quy) [inline language/en-GB/en-GB.mod_banners.ini]: lower case `impressions`

File: language/en-GB/en-GB.mod_banners.ini [hunk 4]
```diff
@@ -19,6 +19,8 @@ MOD_BANNERS_FIELD_HEADER_DESC="Text or HTML to display before the group of banne
 MOD_BANNERS_FIELD_HEADER_LABEL="Header Text"
 MOD_BANNERS_FIELD_RANDOMISE_DESC="Randomise the ordering of the banners."
 MOD_BANNERS_FIELD_RANDOMISE_LABEL="Randomise"
+MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_LABEL="Robots Increment Impressions"
+MOD_BANNERS_FIELD_ROBOTS_VIEWS_DOIMPRESS_DESC="Search engines increase the number of banner impressions"
```
Comment (SharkyKZ) [inline language/en-GB/en-GB.mod_banners.ini]: Add period at the end of the sentence.

**Comment by brianteeman** [comment]
Can someone stop to think if this wording accurately describes what it is intended to do. For example there is no mention that this is "excluding search engine generated 'clicks' from the hit count". It just says it excludes robots which is not technically accurate.

Finally I am still not convinced that this is a useful addition to the core of joomla and is better placed as an extension. If this is accepted then we should have the same option everywhere that we record a "hit"

**Comment by jurihahn** [comment]
Sorry, i think this is not the same. For the banners it's more important. If i see statistics of one of my sites:
real user make about 3 000 page views/impressions per day
search engines make about 13 000 !!! And all can be excluded... 
If i want sell views/impressions, i can't..

Why can't we make it more accurate? This is only an option...

Make better wording is always welcome.

and maybe i do not understand all what you say, but here is nothing to exclude "click", only views/impressions

**Comment by jurihahn** [comment]
> Finally I am still not convinced that this is a useful addition to the core of joomla and is better placed as an extension. If this is accepted then we should have the same option everywhere that we record a "hit"

Really? Why joomla have this in the core? Why we should create new extension for only one option? Joomla banners component is not perfect, but we can make it better. 
It is Joomla politic? Make Joomla extensions not good and say "install third party extension they work better".

**Comment by infograf768** [comment]
The wording is indeed confusing. The feature, in its simplicity, looks OK to me. I see no reason to refuse it.

**Comment by jurihahn** [comment]
> The wording is indeed confusing.

I will be glad to change the text to more suitable. Someone will advise more suitable texts?
Maybe like this.
Label: Search engines increase banner impressions count
Description: If set to NO, detected search engines would not increase impressions count of banner.

**Comment by infograf768** [comment]
I must say I did not test the code and results, though. Therefore it needs testing as I work local.

**Comment by SharkyKZ** [comment]
Regarding usability, I can't think of a scenario where user would want this in one module but not in another. Perhaps this should be set globally in `com_banners` instead of individually in each banner module?

**Comment by jurihahn** [comment]
> I must say I did not test the code and results, though. Therefore it needs testing as I work local.

Its can be tested local too. For example in Firefox you can open Developer tools and start Responsive/Mobile view, then add new Device with any user agent.


> Regarding usability, I can't think of a scenario where user would want this in one module but not in another. Perhaps this should be set globally in `com_banners` instead of individually in each banner module?

yes, I agree.

File: administrator/components/com_banners/config.xml [hunk 5]
```diff
@@ -32,6 +32,18 @@
 			<option value="1">JYES</option>
 			<option value="0">JNO</option>
 		</field>
+
```
Comment (Quy) [inline administrator/components/com_banners/config.xml]: Remove tabs

File: modules/mod_banners/helper.php [hunk 6]
```diff
@@ -32,6 +33,7 @@ public static function &getList(&$params)
 		$document = JFactory::getDocument();
 		$app      = JFactory::getApplication();
 		$keywords = explode(',', $document->getMetaData('keywords'));
+		$config = ComponentHelper::getParams('com_banners');
```
Comment (Quy) [inline modules/mod_banners/helper.php]: Align `=`

**Comment by jurihahn** [comment]
I have tested this PR with #22368 it show very fine result. ~700 impressions per hour before and ~250 impressions per hour after patch.

Maybe i commit one more "option" to ignore click made by robots by the same way? Or should it be different PR?

**Comment by infograf768** [comment]
Maybe for the strings:
 COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_LABEL="Impressions by Search Engines" 
 COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_DESC="If set to NO, prevents search engines clicks from increasing the number of impressions."

**Comment by Quy** [comment]
Be sure to sort the keys in alpha order.

**Comment by jurihahn** [comment]
> Be sure to sort the keys in alpha order.

I don't understand where is it wrong?

**Comment by jurihahn** [comment]
> COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_LABEL="Impressions by Search Engines"

This is good!


> COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_DESC="If set to NO, prevents search engines clicks from increasing the number of impressions."

in this PR is only impressions ignoring, not clicks

File: administrator/language/en-GB/en-GB.com_banners.ini [hunk 7]
```diff
@@ -137,6 +137,8 @@ COM_BANNERS_FIELD_TRACKCLICK_DESC="Record the number of clicks on the banners on
 COM_BANNERS_FIELD_TRACKCLICK_LABEL="Track Clicks"
 COM_BANNERS_FIELD_TRACKIMPRESSION_DESC="Record the impressions (views) of the banners on a daily basis."
 COM_BANNERS_FIELD_TRACKIMPRESSION_LABEL="Track Impressions"
+COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_LABEL="Impressions by Search Engines"
+COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_DESC="If set to NO, prevents search engines from increasing the number of impressions."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_banners.ini]: Move line 141 above line 140.

Comment (jurihahn) [reply administrator/language/en-GB/en-GB.com_banners.ini]: now I understand)
I just watched the new lines were in the right place )))

**Comment by jurihahn** [comment]
@brianteeman what do you think now?

File: administrator/language/en-GB/en-GB.com_banners.ini (hunk 7 above)
Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_banners.ini]: ==>
"Include search engines in the count of impressions"

Comment (jurihahn) [reply administrator/language/en-GB/en-GB.com_banners.ini]: > ==>
> "Include search engines in the count of impressions"

thanks, it's done!

File: administrator/language/en-GB/en-GB.com_banners.ini [hunk 8]
```diff
@@ -137,7 +137,7 @@ COM_BANNERS_FIELD_TRACKCLICK_DESC="Record the number of clicks on the banners on
 COM_BANNERS_FIELD_TRACKCLICK_LABEL="Track Clicks"
 COM_BANNERS_FIELD_TRACKIMPRESSION_DESC="Record the impressions (views) of the banners on a daily basis."
 COM_BANNERS_FIELD_TRACKIMPRESSION_LABEL="Track Impressions"
-COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_DESC="If set to NO, prevents search engines from increasing the number of impressions."
+COM_BANNERS_FIELD_TRACKROBOTSIMPRESSION_DESC="Include search engines in the count of impressions"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_banners.ini]: Add a period.

**Review by brianteeman — APPROVED** [review]
strings are ok

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 2f36648c16c4d3466ba2671eef841089482b20cc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22339">issues.joomla.org/tracker/joomla-cms/22339</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 2f36648c16c4d3466ba2671eef841089482b20cc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22339">issues.joomla.org/tracker/joomla-cms/22339</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22339">issues.joomla.org/tracker/joomla-cms/22339</a>.</sub>

**Pull request merged**